Docs for CLP Pointer permissions. (#700)
ParsePlatform/Docs: _includes/parse-server/class-level-permissions.md

Allowed entries for operations are:

* `*`  -  [Public](#public-access)
* `objectId` - [User's ID](#users-roles)
* `role:role_name` - [Role](#users-roles)
* `requiredAuthentication` - [Authenticated Users](#requires-authentication-permission-requires-parse-server--230)
* `pointerFields` - [Pointer Permissions](#pointer-permissions)

And any combinations of the above:

```js
// PUT http://localhost:1337/schemas/:className
// Set the X-Parse-Application-Id and X-Parse-Master-Key header
// body:
{
  classLevelPermissions:
  {
    "get": {
      "*": true, // means Public access
      "s0meUs3r1d": true, // key must be an id of `_User`
      "role:admin": true, // key must be `role:${role_name}`
      "requiresAuthentication": true, // any authenticated users
      "pointerFields": ["onwer", "followers"] // pointer permissions
    },
    "find": {...},    // ... create, update, delete, addField
    "readUserFields": ["followers", "owner"], // grouped pointer permissions
    "writeUserFields": ["owner"], // grouped pointer permissions
    "protectedFields": {
      "*": ["password","email"],
       "userField:owner":[]
    }
  }
}
```

## Configure CLP Operations

### Public access

> `*` - Allows anyone despite authentication status to execute operation.

Reviewer: ```suggestion
`*` - Allows anyone regardless of authentication status to execute the operation.
```

```js
{
  classLevelPermissions:
  {
    "get": {
      "*": true, // Public access
    }
  }
}
```

### Users, Roles

> This works exactly as ACL's

Reviewer: ```suggestion
This works exactly like ACL's.
```

```js
{
  classLevelPermissions:
  {
    "find": {
      "s0meUs3r1d": true, // key must be an id of `_User`
      "role:admin": true, // key must be `role:${role_name}`
    }
  }
}
```

### Requires Authentication permission

Starting version 2.3.0, Parse Server introduces a new Class Level Permission called `requiresAuthentication`. This CLP prevents any non authenticated user from performing the action protected by the CLP.

> If you want to restrict access to a full class to only authenticated users, you can use the `requiresAuthentication` Class Level Permission. For example, you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you would set the following CLP:

Reviewer: ```suggestion
Starting with version 2.3.0, Parse Server introduced a new Class Level Permission `requiresAuthentication`. It prevents any non authenticated user from performing the action protected by the CLP.

For example, if you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you could set the following CLP:
```

// body:
{
  classLevelPermissions:
  {
    "get": {
      "*": true, // means Public access
      "s0meUs3r1d": true, // key must be an id of `_User`
      "role:admin": true, // key must be `role:${role_name}`
      "requiresAuthentication": true, // any authenticated users
      "pointerFields": ["onwer", "followers"] // pointer permissions
    },
    "find": {...},    // ... create, update, delete, addField
    "readUserFields": ["followers", "owner"], // grouped pointer permissions
    "writeUserFields": ["owner"], // grouped pointer permissions
    "protectedFields": {
      "*": ["password","email"],
       "userField:owner":[]
    }
  }
}
```

## Configure CLP Operations

### Public access

`*` - Allows anyone despite authentication status to execute operation.

```js
{
  classLevelPermissions:
  {
    "get": {
      "*": true, // Public access
    }
  }
}
```

### Users, Roles

This works exactly as ACL's

```js
{
  classLevelPermissions:
  {
    "find": {
      "s0meUs3r1d": true, // key must be an id of `_User`
      "role:admin": true, // key must be `role:${role_name}`
    }
  }
}
```

### Requires Authentication permission

> Starting version 2.3.0, parse-server introduces a new Class Level Permission `requiresAuthentication`.

Reviewer: ```suggestion
Starting version 2.3.0, Parse Server introduces a new Class Level Permission called `requiresAuthentication`. This CLP prevents any non authenticated user from performing the action protected by the CLP.
```

> This CLP prevents any non authenticated user from performing the action protected by the CLP.

Reviewer: ```suggestion
```

> If you want to restrict access to a full class to only authenticated users, you can use the `requiresAuthentication` class level permission. For example, you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you would set the CLP:

Reviewer: ```suggestion
If you want to restrict access to a full class to only authenticated users, you can use the `requiresAuthentication` Class Level Permission. For example, you want to allow your **authenticated users** to `find` and `get` objects from your application and your admin users to have all privileges, you would set the following CLP:
```

}
```

Effects:

* Non authenticated users won't be able to do anything.
* Authenticated users (any user with a valid sessionToken) will be able to read all the objects in that class
* Users belonging to the admin role, will be able to perform all operations.

> ⚠️ Note that this is in no way securing your content, if you allow anyone to log in to your server, any client will still be able to query this object.

Reviewer: ```suggestion
Effects:

* Non authenticated users won't be able to do anything.
* Authenticated users (any user with a valid `sessionToken`) will be able to read all the objects in that class.
* Users belonging to the admin role, will be able to perform all operations.

⚠️ Note that this is in no way securing your content, if you allow anyone to login to your server, every client will still be able to query this object.
```

## Pointer Permissions

While permissions discussed before let you explicitly target a user by id or a role,
Pointer Permissions let you dynamically enforce permissions without knowing the id or assigning roles in advance. Instead you define one or more field names of this class, and any users pointed by such fields of a particular object are granted the permission.

To use this feature, a field must:

* already exist in this collection's schema
* be of either `Pointer<_User>` or `Array` type

In case of `Array`, only items that are of `Pointer<_User>` type are evaluated, other items silently ignored.

You can think of it as a virtual ACL or a dynamic role defined per-object in its own field.

There are two ways you can set Pointer Permission in schema:

* [Using granular permissions](#granular-pointer-permissions) - `pointerFields` *requires Parse-Server v3.11 and above*
* [Using grouped permissions](#grouped-pointer-permissions) - `readUserFields`/`writeUserFields`

⚠️ `create` operation can't be allowed by pointer permissions, because there is literally no object to check it's field untill it is created;

> ⚠️ `addField`  grants permission to only update an object with a new field, but it is advised to set addField permission using other means (e.g. restrict to a role or particular admin user by id).

Reviewer: There is quite a discrepancy between here and the corresponding common section...
https://github.com/parse-community/docs/pull/700/files#diff-22c65bfd6614d949ec20e9d4461ce3cfR163-R169

This part needs some of the grammatical improvements that I suggested in the common section although I wouldn't add all the explanation - in fact I would suggest removing some of the long example paragraphs from the common doc on this section as they seem a bit repetitive / overly wordy.

Author: This brought me to a question - why do we have CLP section in Parse-Server guide in the first place?  It seems to have info about server as a binary, its setup and launch, while CLP is more about schema level feature. I suggest to remove CLP from server guide as a whole.

Reviewer: good question, I think the Parse Server guide has been under loved so only contains limited info about basic / required setup - and there are discrepancies in what is included in the Parse Server README and the guide - I did aim to consolidate all the info into the guide but didn't get around to it.

I'm wondering whether we could remove the CLP section in the server guide in favour of adding the common security doc but I'm not sure how this would work in practice with the code sample etc?

Author: In fact the code samples are mostly schema configs - plain JSON objects. And a few examples come with several demo setup lines (that i've already translated to most sdk languages in this pr). 
Anyway clp and protected fields are used by database admin rather than client code or any of SDKs. I also think it could be very helpful and time saving if we maintain such examples as a shared postman collection.

Reviewer: I see that.

I think adding the common security doc to the server guide would be good, that way there is only one 'copy' to be maintained and its available on all the guides which would encourage devs to think about security when they are implementing their client side.

Reviewer: So are you happy with that or do you have other ideas?

I think it needs to be in there server guide in some way as it is related to setup - especially as you have gone to all the effort of writing it and I don't think it would make sense to have it in the client guides but not the server one?

Author: I agree we should merge it and only have a single copy as you proposed. As for Parse-Server Guide probably REST-flavored examples should fit best.

Reviewer: Cool, REST example would make sense. I'll leave that with you then, or if you've got maintainer commits turned on I could take a look later.

Author: I believe i created this pr with edits enabled. Not sure if there is anything else i should allow?
<img width="261" alt="image" src="https://user-images.githubusercontent.com/2188841/75628552-3c941000-5beb-11ea-9d89-d98ef63d0181.png">
 I'm planning to work on it later today anyway

Reviewer: Yes that’s the one, the annoying thing is that on GitHub there is no indication to me whether you have that ticked or not.

### Granular Pointer Permissions

Given an example setup:

```js
{
  "classLevelPermissions":
  {
    "get": {
      "pointerFields": ["owner", "subscribers"]
    },
    "find": {
      "pointerFields": ["owner", "subscribers"]
    },
    "create":{
      "*": true
    },
    "update": {
      "pointerFields": ["owner"]
    },
    "delete": {
      "pointerFields": ["owner"]
    }
  }
}
```

```js
// Two users:
let alice = ... // Parse.User
let bob = ... // Parse.User

// and two objects:
let feedA = {
  title: 'Posts by Alice',
  owner: alice,
  subscribers: [], // notice no subscribers here
};

let feedB = {
  title: 'Posts by Bob',
  owner: bob,
  subscribers: [alice], // notice Alice in subscribers
};
```

In the example above:

* anyone is allowed to `create` objects.
* `feed1` can be viewed (`get`,`find`) only by **Alice**. (pointed in owners field)
* `feed2` can be viewed (`get`,`find`) both by **Bob** and **Alice**. (Bob is pointed in `owners`. Alice is pointed in `subscribers`)

> * only owners are allowed to `update` and `delete`.

Reviewer: ```suggestion
In the example above:

* Anyone is allowed to `create` objects.
* `feedA` can be viewed (`get`,`find`) only by **Alice** (pointed to in `owners` field).
* `feedB` can be viewed (`get`,`find`) both by **Bob** and **Alice** (Bob is pointed to in `owners`, Alice is pointed to in `subscribers`).
* Only owners are allowed to `update` and `delete`.
```

### Grouped Pointer Permissions

These are similar to [`pointerFields`](#granular-pointer-permissions), but cover multiple operations at once:

**`readUserFields`**:

* `get`,
* `find`,
* `count`

**`writeUserFields`**:

* `update`,
* `delete`,
* `addField`

> Same scheme as for previous example can be defined shorter:

Reviewer: ```suggestion
**`readUserFields`**:

* `get`
* `find`
* `count`

**`writeUserFields`**:

* `update`
* `delete`
* `addField`

The same scheme as for previous example can be defined in a more compact manner:
```

Now lets look at example that uses Pointer Permissions. Say we have a `Post` class, with an object, `myPost`. There are 2 users in our app, `poster`, and `viewer`. Lets say we add a pointer permission that gives anyone in the `Creator` field of the `Post` class read and write access to the object, and for the `myPost` object, `poster` is the user in that field. There is also an ACL on the object that gives read access to `viewer`. You may expect that this will allow `poster` to read and edit `myPost`, and `viewer` to read it, but `viewer` will be rejected by the Pointer Permission, and `poster` will be rejected by the ACL, so again, neither user will be able to access the object.

> Because of the complex interaction between CLPs, Pointer Permissions, and ACLs, we recommend being careful when using them together. Often it can be useful to use CLPs only to disable all permissions for a certain request type, and then using Pointer Permissions or ACLs for other request types. For example, you may want to disable Delete for a `Photo` class, but then put a Pointer Permission on `Photo` so the user who created it can edit it, just not delete it. Because of the especially complex way that Pointer Permissions and ACLs interact, we usually recommend only using one of those two types of security mechanisms.

Reviewer: I'll look at this properly later but at first glance this seems very dense. Would it be possible to shorten this section or replace some of it with a diagram or schema.

Author: agree, I believe there is some room to make it more concise. May be code will read easier than words
